### ISSUE TITLE
Segmentation Fault on Power 8 system (Summit)

### ISSUE THREAD

**Issue by csadorf** [glotzerlab/freud#246]
**Brief description**

Executing unit tests and computing a radial distribution function causes a Segmentation Fault on [INCITE Summit](https://www.olcf.ornl.gov/for-users/system-user-guides/summit).

**To Reproduce**

1. Install freud into an environment on Incite Summit with the following module environment:

```
$ module list
Currently Loaded Modules:
  1) hsi/5.0.2.p5   3) lsf-tools/2.0           5) DefApps     7) spectrum-mpi/10.2.0.11-20190201   9) cmake/3.13.4  11) netlib-lapack/3.8.0           13) hdf5/1.10.3
  2) xalt/1.1.3     4) darshan-runtime/3.1.7   6) gcc/6.4.0   8) cuda/9.2.148                     10) git/2.13.0    12) python/3.6.6-anaconda3-5.3.0
```
and this command:

```
python setup.py install
```
2. Then execute:

```
cd tests
python -m unittest discover -v
```

**Error output**

```
(env-summit) [csadorf@login1.summit tests]$ python -m unittest discover . -v
test_2_dimensional (test_box_Box.TestBox) ... ok
test_box_volume (test_box_Box.TestBox) ... ok
test_construct (test_box_Box.TestBox)
Test correct behavior for various constructor signatures ... ok
test_coordinates (test_box_Box.TestBox) ... ERROR
test_cube (test_box_Box.TestBox) ... ok
test_equal (test_box_Box.TestBox) ... ok
test_fraction (test_box_Box.TestBox) ... ERROR
test_from_box (test_box_Box.TestBox)
Test various methods of initializing a box ... ok
test_get_length (test_box_Box.TestBox) ... ok
test_get_tilt_factor (test_box_Box.TestBox) ... ok
test_images (test_box_Box.TestBox) ... ERROR
test_matrix (test_box_Box.TestBox) ... ok
test_multiply (test_box_Box.TestBox) ... ok
test_periodic (test_box_Box.TestBox) ... ok
test_set_length (test_box_Box.TestBox) ... ok
test_square (test_box_Box.TestBox) ... ok
test_str (test_box_Box.TestBox) ... ok
test_to_dict (test_box_Box.TestBox)
Test converting box to dict ... ok
test_tuple (test_box_Box.TestBox) ... ok
test_unwrap (test_box_Box.TestBox) ... ok
test_vectors (test_box_Box.TestBox)
Test getting lattice vectors ... ok
test_wrap_multiple_images (test_box_Box.TestBox) ... ok
test_wrap_multiple_particles (test_box_Box.TestBox) ... ok
test_wrap_single_particle (test_box_Box.TestBox) ... ok
test_cube (test_box_ParticleBuffer.TestParticleBuffer) ... FAIL
test_square (test_box_ParticleBuffer.TestParticleBuffer) ... FAIL
test_triclinic (test_box_ParticleBuffer.TestParticleBuffer) ... FAIL
test_cluster_keys (test_cluster.TestCluster) ... Segmentation fault (core dumped)
```

Traceback in combination with gdb:

```
HOOMD-blue is running on the following GPU(s):
[New Thread 0x20003050f190 (LWP 94571)]
[New Thread 0x20003091f190 (LWP 94572)]
259     cpp/density/RDF.cc: No such file or directory.
[New Thread 0x2000312cf190 (LWP 94576)]
[New Thread 0x2000316df190 (LWP 94577)]
[New Thread 0x200031aef190 (LWP 94578)]
[New Thread 0x200031eff190 (LWP 94579)]
[New Thread 0x20003230f190 (LWP 94580)]
[New Thread 0x20003271f190 (LWP 94581)]
[New Thread 0x200032b2f190 (LWP 94582)]
[New Thread 0x20003375f190 (LWP 94584)]
[New Thread 0x200032f3f190 (LWP 94583)]
[New Thread 0x20003e4df190 (LWP 94589)]
[New Thread 0x200033f7f190 (LWP 94587)]
[New Thread 0x200033b6f190 (LWP 94586)]
[New Thread 0x20003334f190 (LWP 94585)]
[New Thread 0x20003e0cf190 (LWP 94588)]
[New Thread 0x20003ecff190 (LWP 94591)]
[New Thread 0x20003e8ef190 (LWP 94590)]
[New Thread 0x20003f10f190 (LWP 94593)]
[New Thread 0x20003f51f190 (LWP 94594)]
[New Thread 0x20003f92f190 (LWP 94595)]
[New Thread 0x20003fd3f190 (LWP 94596)]
[New Thread 0x2000a881f190 (LWP 94597)]
[New Thread 0x2000a440f190 (LWP 94598)]
[New Thread 0x2000a481f190 (LWP 94599)]
[New Thread 0x2000a4c2f190 (LWP 94600)]
[New Thread 0x2000a503f190 (LWP 94601)]
[New Thread 0x2000a544f190 (LWP 94602)]
[New Thread 0x2000a585f190 (LWP 94603)]

Program received signal SIGSEGV, Segmentation fault.
operator() (r=<optimized out>, __closure=0x200030ea3860)
    at cpp/density/RDF.cc:259
Missing separate debuginfos, use: debuginfo-install glibc-2.17-222.el7.ppc64le keyutils-libs-1.5.8-3.el7.ppc64le krb5-libs-1.15.1-19.el7.ppc64le libcom_err-1.42.9-12.el7_5.ppc64le libibverbs-41mlnx1-OFED.4.3.2.1.6.43401.0.ppc64le libmlx4-41mlnx1-OFED.4.1.0.1.0.43401.0.ppc64le libmlx5-41mlnx1-OFED.4.3.4.0.3.43401.0.ppc64le libnl3-3.2.28-4.el7.ppc64le librxe-41mlnx1-OFED.4.4.2.4.6.43401.0.ppc64le libselinux-2.5-12.el7.ppc64le numactl-libs-2.0.9-7.el7.ppc64le openssl-libs-1.0.2k-12.el7.ppc64le pcre-8.32-17.el7.ppc64le xz-libs-5.2.2-1.el7.ppc64le
(gdb) User defined signal 2
Could not read jskill result from pmix server
```

**System configuration**:

 - OS: `4.14.0-49.18.1.el7a.ppc64le` (output of `uname -r`)
 - Version of Python: 3.7
 - Version of freud: 1.0
 - [freud-issue-246-environment.txt](https://github.com/glotzerlab/freud/files/3076679/freud-issue-246-environment.txt)

**Additional context**

 - Issues occur both on the login and the compute nodes.
 - Issues do **not** occur in other environments with an identical version of freud and the **identical** data set!

**Comment by bdice** [comment]
I was able to reproduce this issue. It seems that building with gcc 8.1.1 stops the segfaults.

Using gcc 8.1.1 requires something like this:
```bash
module load gcc/8.1.1
CC=$(which gcc) CXX=$(which g++) python setup.py install --ENABLE-CYTHON
```

When I try to run the unit tests after building with gcc 8.1.1, some fail due to numerical precision (I will make a PR reducing the precision of the checks for those tests). There is still some issue with the Nematic Order Parameter test. I have no idea if this is related to the segfault issues or not:
```python
>>> n = freud.order.NematicOrderParameter([1, 0, 0])
>>> n
<freud.order.NematicOrderParameter object at 0x7fff747aa060>
>>> n.director
array([0., 0., 0.], dtype=float32)
```

**Comment by csadorf** [comment]
I've recompiled freud with `gcc/7.4.0` and I don't get any segmentation faults anymore, however some of the unit tests fail.

```
$ module list

Currently Loaded Modules:
  1) hsi/5.0.2.p5   3) lsf-tools/2.0           5) DefApps                        7) gcc/7.4.0                         9) cuda/9.2.148  11) git/2.13.0           13) hdf5/1.10.3
  2) xalt/1.1.3     4) darshan-runtime/3.1.7   6) python/3.7.0-anaconda3-5.3.0   8) spectrum-mpi/10.2.0.11-20190201  10) cmake/3.13.4  12) netlib-lapack/3.8.0
```

[unittest_output.txt](https://github.com/glotzerlab/freud/files/3098806/unittest_output.txt)

**Comment by vyasr** [comment]
Thanks for reporting! That makes sense, we realized recently that our tests are testing for unreasonable precisions, we'll need to fix that. Glad the compiler fixes your primary issue though.

**Comment by csadorf** [comment]
I don't think that all of the failed tests are related to precision issues, but I'll leave that up to you to analyze.

**Comment by bdice** [comment]
After #254, I believe the unit tests should all pass when compiled with `gcc` >= 7. The problems with `gcc` 6 are unexpected and hard to identify -- this issue can be re-opened if problems persist.

**Comment by balewski** [comment]
Hi,
I'm running TF on Summit and see the same error 
- - - 
Could not read jskill result from pmix server
User defined signal 2
- - - 
but only for some jobs. It is not reproducible. Is there any cure for that?
Jan

**Comment by bdice** [comment]
@balewski Are you using the `freud` library with TensorFlow? If so, please file a new issue with the full error log, with the version of `freud` and which compilers you're using. If this is only about TensorFlow on Summit and not `freud`, you may wish to ask for help from OLCF support. For the segmentation faults in `freud` that were mentioned in this issue, using a newer compiler (gcc >= 7) helped resolve the issue.
